File: backend/ml/tracing/tracing.py
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.sdk.resources import Resource
import os
import logging

logger = logging.getLogger(__name__)

class Tracing:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance
    
    def __init__(self):
        self.provider = None
        self.is_initialized = False
    
    def initialize(self, service_name="truxify-ml"):
        if self.is_initialized:
            return
        
        try:
            # Create resource
            resource = Resource.create({
                "service.name": service_name,
                "service.version": "1.0.0",
                "deployment.environment": os.getenv("ENVIRONMENT", "development"),
            })
            
            # Create provider
            self.provider = TracerProvider(resource=resource)
            
            # Add span processor
            otlp_exporter = OTLPSpanExporter(
                endpoint=os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4317"),
                insecure=True,
            )
            self.provider.add_span_processor(
                BatchSpanProcessor(otlp_exporter)
            )
            
            # Set provider
            trace.set_tracer_provider(self.provider)
            
            self.is_initialized = True
            logger.info(f"✅ OpenTelemetry initialized for {service_name}")
            
        except Exception as e:
            logger.error(f"❌ OpenTelemetry initialization failed: {e}")
    
    def instrument_fastapi(self, app):
        """Instrument FastAPI application"""
        if self.is_initialized:
            FastAPIInstrumentor.instrument_app(
                app,
                tracer_provider=self.provider,
                excluded_urls="/health,/metrics"
            )
            logger.info("✅ FastAPI instrumented")
    
    def instrument_requests(self):
        """Instrument requests library"""
        if self.is_initialized:
            RequestsInstrumentor().instrument(tracer_provider=self.provider)
            logger.info("✅ Requests instrumented")
    
    def instrument_sqlalchemy(self, engine):
        """Instrument SQLAlchemy"""
        if self.is_initialized:
            SQLAlchemyInstrumentor().instrument(
                engine=engine,
                tracer_provider=self.provider
            )
            logger.info("✅ SQLAlchemy instrumented")
    
    def get_tracer(self, name="truxify-ml"):
        if not self.is_initialized:
            self.initialize()
        return trace.get_tracer(name)
    
    def create_span(self, name):
        tracer = self.get_tracer()
        return tracer.start_span(name)
    
    def shutdown(self):
        if self.provider:
            self.provider.shutdown()

tracing = Tracing()

File: backend/api/src/tracing/tracing.js
import { NodeTracerProvider } from '@opentelemetry/sdk-trace-node';
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base';
import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { registerInstrumentations } from '@opentelemetry/instrumentation';
import { PinoInstrumentation } from '@opentelemetry/instrumentation-pino';
import { MongoDBInstrumentation } from '@opentelemetry/instrumentation-mongodb';
import { RedisInstrumentation } from '@opentelemetry/instrumentation-redis';
import { WinstonInstrumentation } from '@opentelemetry/instrumentation-winston';
import logger from '../middleware/logger.js';

class Tracing {
    constructor() {
        this.provider = null;
        this.isInitialized = false;
    }

    initialize(serviceName = 'truxify-api') {
        if (this.isInitialized) return;

        try {
            // Create resource
            const resource = new Resource({
                [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
                [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
                [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: process.env.NODE_ENV || 'development',
                [SemanticResourceAttributes.HOST_NAME]: process.env.HOSTNAME || 'localhost',
            });

            // Create provider
            this.provider = new NodeTracerProvider({
                resource: resource,
                spanProcessors: [
                    new BatchSpanProcessor(
                        new OTLPTraceExporter({
                            url: process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://localhost:4317',
                            timeoutMillis: 10000,
                        })
                    )
                ]
            });

            // Register provider
            this.provider.register();

            // Register instrumentations
            this.registerInstrumentations();

            this.isInitialized = true;
            logger.info(`✅ OpenTelemetry initialized for ${serviceName}`);
        } catch (error) {
            logger.error('❌ OpenTelemetry initialization failed:', error);
        }
    }

    registerInstrumentations() {
        registerInstrumentations({
            instrumentations: [
                new ExpressInstrumentation({
                    ignoreLruCache: true,
                    enabled: true,
                }),
                new HttpInstrumentation({
                    ignoreIncomingPaths: ['/health', '/metrics', '/favicon.ico'],
                    enabled: true,
                }),
                new PinoInstrumentation({
                    enabled: true,
                }),
                new WinstonInstrumentation({
                    enabled: true,
                }),
                new MongoDBInstrumentation({
                    enabled: true,
                }),
                new RedisInstrumentation({
                    enabled: true,
                }),
            ]
        });
    }

    getTracer(name = 'truxify') {
        if (!this.isInitialized) {
            this.initialize();
        }
        return this.provider.getTracer(name);
    }

    createSpan(name, options = {}) {
        const tracer = this.getTracer();
        return tracer.startSpan(name, options);
    }

    async withSpan(name, fn, options = {}) {
        const tracer = this.getTracer();
        const span = tracer.startSpan(name, options);
        
        try {
            const result = await fn(span);
            span.end();
            return result;
        } catch (error) {
            span.recordException(error);
            span.setStatus({ code: 2, message: error.message });
            span.end();
            throw error;
        }
    }

    addAttributes(span, attributes) {
        if (span) {
            span.setAttributes(attributes);
        }
    }

    addEvent(span, name, attributes = {}) {
        if (span) {
            span.addEvent(name, attributes);
        }
    }

    getActiveSpan() {
        return this.provider?.getActiveSpan();
    }

    shutdown() {
        if (this.provider) {
            return this.provider.shutdown();
        }
    }
}

export default new Tracing();